### PR TITLE
support 1.24

### PR DESCRIPTION
**Pull request by jiaqiluo**
**Issues**: 
- https://github.com/rancher/rke/issues/2936
- https://github.com/rancher/rke/issues/2938

**Changes in this PR**:
- update related Go packages to the 1.24 version
- enforce `cri-dockerd` to be enabled when the cluster version >= 1.24 and the option `enable_cri_dockerd` is not configured in the cluster.yml file

**Test**:
This PR has been tested locally by pointing the KDM to my forked which contains the template for `1.24.1-rancher1-1`.  
- The provision of a single-node `1.24.1-rancher1-1` cluster succeeded. 
- The provision of a single-node `1.24.1-rancher1-1` cluster failed as expected when `enable_cri_dockerd=false` is set in the cluster.yml file
- Upgrading a single-node cluster from `v1.23.6-rancher1-1` to `1.24.1-rancher1-1` also succeeded. 

**Related PRs**:
- https://github.com/rancher/kontainer-driver-metadata/pull/910

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -33,12 +33,12 @@ require (
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 	google.golang.org/grpc v1.40.0
 	gopkg.in/yaml.v2 v2.4.0
-	k8s.io/api v0.23.3
-	k8s.io/apimachinery v0.23.3
-	k8s.io/apiserver v0.23.3
-	k8s.io/client-go v0.23.3
-	k8s.io/gengo v0.0.0-20210813121822-485abfe95c7c
-	k8s.io/kubectl v0.23.3
+	k8s.io/api v0.24.1
```
Comment (kinarashah) [inline go.mod]: could you update this now to 0.24.2

File: cluster/plan.go [hunk 2]
```diff
@@ -345,7 +349,9 @@ func (c *Cluster) BuildKubeControllerProcess(host *hosts.Host, serviceOptions v3
 	}
 	CommandArrayArgs := make(map[string][]string, len(c.Services.KubeAPI.ExtraArgsArray))
 	// Best security practice is to listen on localhost, but DinD uses private container network instead of Host.
-	if c.DinD {
+	// the flag --address is removed since k8s 1.24
+	parsedVersion, _ := getClusterVersion(c.Version)
```
Comment (kinarashah) [inline cluster/plan.go]: need to handle err here

Comment (kinarashah) [reply cluster/plan.go]: ok if we don't return, but we can log it

**Review by kinarashah — APPROVED** [review]

File: cluster/defaults.go [hunk 3]
```diff
@@ -288,6 +294,23 @@ func (c *Cluster) setNodeUpgradeStrategy() {
...
 }
 
+// setCRIDockerd set enable_cri_dockerd = true when the following two conditions are met: the cluster's version is
+// at least 1.24 and the option enable_cri_dockerd is not configured
+func (c *Cluster) setCRIDockerd() error {
+	parsedVersion, err := getClusterVersion(c.Version)
+	if err != nil {
+		return err
+	}
+	r := semver.MustParseRange(">= 1.24.0-rancher0")
```
Comment (kinarashah) [inline cluster/defaults.go]: nit: could you use var `parsedRangeAtLeast124` already declared in cluster/plan.go

File: go.mod (hunk 1 above)
Comment (jiaqiluo) [reply go.mod]: Yep, updated related packages to 0.24.2

File: cluster/plan.go (hunk 2 above)
Comment (jiaqiluo) [reply cluster/plan.go]: fixed

File: cluster/defaults.go (hunk 3 above)
Comment (jiaqiluo) [reply cluster/defaults.go]: good catch
fixed

**Review by kinarashah — APPROVED** [review]

**Review by rosskirkpat — APPROVED** [review]

**Pull request merged**